packages/workspace/src/workspace/element_selector.ts

```diff
@@ -142,118 +123,93 @@ export const createElementSelectors = (selectors: string[]):
-  } => {
-  const [topLevelSelectors, subElementSelectors] = _.partition(selectors, selector =>
-    isTopLevelSelector(selector, depth))
-  const { elements: currentLevelElementsSelected, matches } = topLevelSelectors.length > 0
-    ? selectElementsBySelectors(elements, topLevelSelectors, false)
-    : { elements: [], matches: {} }
-  return {
-    subElementSelectors,
-    matches,
-    currentLevelElementsSelected:
-      compact ? removeChildElements(currentLevelElementsSelected) : currentLevelElementsSelected,
-  }
-}
-
-const getElementIdsFromSelectorsRecursivelyInner = async (
-  selectors: ElementSelector[], elements: ElementIDToValue[], compact = false, depth = 1
-): Promise<ElementIDToValueResult> => {
-  const { subElementSelectors, matches,
-    currentLevelElementsSelected } = selectElementsForDepth(selectors, elements, compact, depth)
-  if (_.isEmpty(subElementSelectors)) {
-    return { elementIdToValue: currentLevelElementsSelected, matches }
-  }
-  const subElementParentSelectors = createParentSelectors(subElementSelectors,
-    currentLevelElementsSelected, compact, depth)
-  if (subElementParentSelectors.length === 0) {
-    return { elementIdToValue: currentLevelElementsSelected, matches }
+export const selectElementIdsByTraversal = async (
+  selectors: ElementSelector[], elements: ElementIDToValue[],
+  compact = false): Promise<ElemID[]> => {
+  const [wildcardSelectors, determinedSelectors] = _.partition(selectors, selector => selector.origin.includes('*'))
+  const ids = determinedSelectors.map(selector => ElemID.fromFullName(selector.origin))
+  if (_.isEmpty(wildcardSelectors)) {
+    return ids
+  }
+  const [topLevelSelectors, subElementSelectors] = _.partition(wildcardSelectors,
+    isTopLevelSelector)
+  if (!_.isEmpty(topLevelSelectors)) {
+    const { elements: topLevelElements } = selectElementsBySelectors(elements,
+      topLevelSelectors, false)
+    ids.push(...topLevelElements.map(element => element.elemID).concat(ids))
+    if (_.isEmpty(subElementSelectors)) {
+      return ids
+    }
   }
-  const { elements: possibleParentElements } = selectElementsBySelectors(elements,
-    subElementParentSelectors, false)
-  if (possibleParentElements.length === 0) {
-    return { elementIdToValue: currentLevelElementsSelected, matches }
+  const possibleParentSelectors = subElementSelectors.map(createTopLevelSelector)
+  const stillRelevantElements = compact ? selectElementsBySelectors(elements,
+    possibleParentSelectors, false).elements.filter(id => !ids.includes(id.elemID))
+    : selectElementsBySelectors(elements, possibleParentSelectors, false).elements
+  if (_.isEmpty(stillRelevantElements)) {
+    return ids
+  }
+  const transformFunc = (args: TransformFuncArgs): Value | undefined => {
+    if (args.path) {
+      const testId = args.path
+      const { elements: found } = selectElementsBySelectors([testId], subElementSelectors, false)
+      if (!_.isEmpty(found)) {
```

Comment:
nit: prefer native language (saw in several places, but you can decide to ignore)
```suggestion
      if (found.length !== 0) {
```